### PR TITLE
[kernel] introduce fused_moe_triton_splitk optimization

### PR DESCRIPTION
**Pull request by BBuf**
## Background

I got the idea from: [blog](https://pytorch.org/blog/accelerating-moe-model) and [code](https://github.com/pytorch-labs/applied-ai/blob/main/kernels/triton/inference/col_major_moe_gemm/v1_moe_fused.py)

PyTorch team claimed that the performance of the `fused_moe_triton_splitk` is better than the original `fused_moe_triton` kernel on [A100 and H100 when deploying mixtral-moe](https://github.com/pytorch-labs/applied-ai/issues/17#issuecomment-2043422080).

Therefore, in this PR, I introduced `fused_moe_triton_splitk` to sglang, which was derived from the original `fused_moe_triton` with some modifications. Due to `atomic_add` operations causing random values when writing to the output tensor, I can't get correct result for a long time , but I resolved the bug through debugging today. I performed development, accuracy testing, and performance validation on a GTX 4090, where I didn't see performance gains. However, I believe that newer GPUs like the A100 or H100 would show much performance improvements as [claimed by Meta](https://github.com/pytorch-labs/applied-ai/issues/17#issuecomment-2043422080).

### fused_moe_triton

serving command:

```shell
python3 -m sglang.launch_server --model-path /mnt/bbuf/benchmark_data/Qwen2-57B-A14B-Instruct-FP8   --context-length 4096  --chunked-prefill-size 512 --schedule-policy lpm --load-balance-method round_robin --enable-mixed-chunk --kv-cache-dtype auto --port 8000 --tp 4 --dp 2 --host 0.0.0.0 --disable-custom-all-reduce --mem-fraction-static 0.8 --disable-cuda-graph --dtype float16 --host 0.0.0.0 
```

eval command:

```shell
lm_eval --model local-completions --tasks lambada_openai,hellaswag --model_args model=/mnt/bbuf/benchmark_data/Qwen2-57B-A14B-Instruct-FP8/,base_url=http://127.0.0.1:8000/v1/completions,num_concurrent=1,max_retries=3,tokenized_requests=False --batch_size 64 --output_path sglang-fp8-output/
```

result:

|    Tasks     |Version|Filter|n-shot|  Metric  |   |Value |   |Stderr|
|--------------|------:|------|-----:|----------|---|-----:|---|-----:|
|hellaswag     |      1|none  |     0|acc       |↑  |0.6513|±  |0.0048|
|              |       |none  |     0|acc_norm  |↑  |0.8396|±  |0.0037|
|lambada_openai|      1|none  |     0|acc       |↑  |0.7607|±  |0.0059|
|              |       |none  |     0|perplexity|↓  |2.9287|±  |0.0574|

### fused_moe_triton_split

serving command:
```shell
SGLANG_FUSED_MOE_BACKEND=GEMM_SPLITK python3 -m sglang.launch_server --model-path /mnt/bbuf/benchmark_data/Qwen2-57B-A14B-Instruct-FP8   --context-length 4096  --chunked-prefill-size 512 --schedule-policy lpm --load-balance-method round_robin --enable-mixed-chunk --kv-cache-dtype auto --port 8000 --tp 4 --dp 2 --host 0.0.0.0 --disable-custom-all-reduce --mem-fraction-static 0.8 --disable-cuda-graph --dtype float16 --host 0.0.0.0 
```

eval command:
```shell
lm_eval --model local-completions --tasks lambada_openai,hellaswag --model_args model=/mnt/bbuf/benchmark_data/Qwen2-57B-A14B-Instruct-FP8/,base_url=http://127.0.0.1:8000/v1/completions,num_concurrent=1,max_retries=3,tokenized_requests=False --batch_size 64 --output_path sglang-fp8-output/
```

result:

|    Tasks     |Version|Filter|n-shot|  Metric  |   |Value |   |Stderr|
|--------------|------:|------|-----:|----------|---|-----:|---|-----:|
|hellaswag     |      1|none  |     0|acc       |↑  |0.6498|±  |0.0048|
|              |       |none  |     0|acc_norm  |↑  |0.8383|±  |0.0037|
|lambada_openai|      1|none  |     0|acc       |↑  |0.7632|±  |0.0059|
|              |       |none  |     0|perplexity|↓  |2.9276|±  |0.0575|

## GTX 4090 Performace

图片

I didn't see performance gains in GTX 4090 with tp4 w8a8 settings in Qwen2-57B model. However, I believe that newer GPUs like the A100 or H100 would show much performance improvements as [claimed by Meta](https://github.com/pytorch-labs/applied-ai/issues/17#issuecomment-2043422080).

## Limitation

It's worth noting that since Triton doesn't support atomic_add operations on BF16, you need to specify dtype as float16 when deploying the model. Please refer https://github.com/triton-lang/triton/pull/2708 .

### CODE REVIEW

File: python/sglang/srt/layers/fused_moe_triton/fused_moe_splitk.py [hunk 1]
```diff
@@ -0,0 +1,867 @@
+# Adapted from https://github.com/pytorch-labs/applied-ai/blob/main/kernels/triton/inference/col_major_moe_gemm/v1_moe_fused.py
```
Comment (BBuf) [inline python/sglang/srt/layers/fused_moe_triton/fused_moe_splitk.py:1]: Initially, I didn't consider putting the splitk implementation in a separate file. However, keeping it in the same file as `fused_moe.py` would require additional conditional logic, and importing common functions from `fused_moe.py` might lead to circular import issues. After consideration, I decided to create a new file.

File: benchmark/kernels/fused_moe_triton/tuning_fused_moe_triton_splitk.py [hunk 2]
```diff
@@ -0,0 +1,455 @@
+# Adapted from https://github.com/vllm-project/vllm/blob/main/benchmarks/kernels/benchmark_moe.py
```
Comment (BBuf) [inline benchmark/kernels/fused_moe_triton/tuning_fused_moe_triton_splitk.py:1]: Almost the same reason with `fused_moe_splitk.y` why I use indepent file with tuning splitk best config.

**Comment by merrymercy** [comment]
Thanks for the contribution. Can you fix the lint error?

**Comment by BBuf** [comment]
> Thanks for the contribution. Can you fix the lint error?

done

**Comment by dhou-xai** [comment]
can you implement this as a flag in the existing fused_moe kernel? then you can configure with the json.

I don't think it is good to enable atomicAdd by default due to nondeterminism. maybe it is possible to sum at the end and fuse with the expert gather sum.

for the flag, make it a SPLIT_K: tl.constexpr, and make the kernel be the same as before if it is 1.

**Comment by yzh119** [comment]
As @dhou-xai mentioned, it's not encouraged to use `atomicAdd` in production for reproducibility.

**Comment by BBuf** [comment]
> can you implement this as a flag in the existing fused_moe kernel? then you can configure with the json.
> 
> I don't think it is good to enable atomicAdd by default due to nondeterminism. maybe it is possible to sum at the end and fuse with the expert gather sum.
> 
> for the flag, make it a SPLIT_K: tl.constexpr, and make the kernel be the same as before if it is 1.



> can you implement this as a flag in the existing fused_moe kernel? then you can configure with the json.
> 
> I don't think it is good to enable atomicAdd by default due to nondeterminism. maybe it is possible to sum at the end and fuse with the expert gather sum.
> 
> for the flag, make it a SPLIT_K: tl.constexpr, and make the kernel be the same as before if it is 1.

I see. If we want to remove `atomic_add`, kernel need additional memory buffer, and after the fused moe computation is complete, an extra reduce sum step will be needed. If this overhead is acceptable, I can make the modifications.

**Comment by BBuf** [comment]
> can you implement this as a flag in the existing fused_moe kernel? then you can configure with the json.
> 
> I don't think it is good to enable atomicAdd by default due to nondeterminism. maybe it is possible to sum at the end and fuse with the expert gather sum.
> 
> for the flag, make it a SPLIT_K: tl.constexpr, and make the kernel be the same as before if it is 1.

Yeah, when splitk=1 we can indeed directly use `tl.store`. Are you suggesting to add the flag in `fused_moe_splitk.py` rather than original `fused_moe.py`? Since I plan to add more fused moe kernel backend options in the future, and making all modifications in `fused_moe.py` could make it difficult to maintain.

**Comment by BBuf** [comment]
I've reconsidered. For `fused_moe_splitk`, if we don't use atomic_add, I would need an additional buffer with shape `[SPLIT_K, num_tokens, hidden_size]` to assist with the summation, and this memory overhead would increase as `SPLIT_K` increases. Therefore, I'm currently inclined to keep the original atomic_add implementation. If anyone has better implementation ideas, they're welcome to help optimize this implementation. Regarding the determinism issue caused by `atomic_add`, I can consider adding a unit test for the operator and throwing a warning when users choose to use `fused_moe_splitk`.

**Comment by yzh119** [comment]
Serial split-k don't need additional buffers and guarantees reproducibility but you have to maintain semaphore in cuda, I don't know if it's easy to do so in triton.

Having additional workspace buffer is okay IMO. The size of the buffer would not increase too much as K increases because it's finally determined by the number of SMs which is not too large.

Cutlass's group gemm implementation also has a workspace buffer argument. You can refer to flashinfer's group gemm API.

**Comment by dhou-xai** [comment]
> I don't know if it's easy to do so in triton

It would be interesting to figure this out.

**Comment by BBuf** [comment]
I have a new idea to 

> Serial split-k don't need additional buffers and guarantees reproducibility but you have to maintain semaphore in cuda, I don't know if it's easy to do so in triton.
> 
> Having additional workspace buffer is okay IMO. The size of the buffer would not increase too much as K increases because it's finally determined by the number of SMs which is not too large.
> 
> Cutlass's group gemm implementation also has a workspace buffer argument. You can refer to flashinfer's group gemm API.

I've investigated and found that it's currently challenging to avoid using atomic_add in Triton. However, I have a better approach to implement optimizations similar to splitk, but it requires using cutlass. I'll close this PR for now, and we can discuss these optimizations after I complete the implementation.

**Pull request closed without merging**